### PR TITLE
Add MAP cast support

### LINKED ISSUE #4821: Error when coercing VARCHAR in MAP<VARCHAR, T>
`select * from (values (map(array['test'],array[1])), (map(ARRAY['test2'], array[1])));`
throws an exception:
`Cannot cast map(varchar(4),bigint) to map(varchar(5),bigint)`

I think we need to implement CAST operator for MAP<T, T> to MAP<T, T> in order to make it work. ARRAY type works as expected because there is ArrayToArrayCast that takes care of coercing inline types.

### PR DESCRIPTION
**Pull request by geraint0923**
Fix #4821

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MapToMapCast.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+            ConnectorSession session,
+            @SqlType("map(FK,FV)") Block fromMap)
+    {
+        TypedSet typedSet = new TypedSet(toKeyType, fromMap.getPositionCount() / 2);
+        BlockBuilder keyBlockBuilder = toKeyType.createBlockBuilder(new BlockBuilderStatus(), fromMap.getPositionCount() / 2);
+        for (int i = 0; i < fromMap.getPositionCount(); i += 2) {
+            Object fromKey = readNativeValue(fromKeyType, fromMap, i);
+            try {
+                Object toKey;
+                if (fromKey instanceof Block) {
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MapToMapCast.java]: Why does this depend on whether `fromKey` is a `Block`?

Comment (geraint0923) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MapToMapCast.java]: If `fromKey` is `Block`, then it should be a compound structure such as ARRAY whose cast method will take two argument one is `ConnectorSession ` and the other is the corresponding `Block`. 

See https://github.com/prestodb/presto/blob/342cee4d430540915bdb2b269f5133588992c328/presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayToArrayCast.java#L87

Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MapToMapCast.java]: That's an incorrect assumption. Any cast can take a session as the first argument (e.g., datetime casts). The right thing to do would be to determine whether the MethodHandle takes a session as the first argument and pass it if necessary. Ideally, the framework should provide a MethodHandle with the first argument already bound, so this code doesn't need to care about it.

Comment (geraint0923) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MapToMapCast.java]: You are right, casting timestamp also requires `ConnectorSession`.
```
select CAST(MAP(ARRAY[1,2], ARRAY[TIMESTAMP '2016-01-02 01:02:03', TIMESTAMP '2016-02-03 03:04:05']) AS MAP(bigint, varchar));
```
will fail with:
```
Query 20160422_061124_00008_2e56p failed: cannot convert MethodHandle(ConnectorSession,long)Slice to (Object)Object
```
I will change to bind the `ConnectorSession` if needed and add more tests to catch such problems.

**Comment by geraint0923** [comment]
Address the comments.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MapToMapCast.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+            @OperatorDependency(operator = CAST, returnType = "TV", argumentTypes = {"FV"}) MethodHandle valueCastFunction,
+            @TypeParameter("FK") Type fromKeyType,
+            @TypeParameter("FV") Type fromValueType,
+            @TypeParameter("TK") Type toKeyType,
+            @TypeParameter("TV") Type toValueType,
+            ConnectorSession session,
+            @SqlType("map(FK,FV)") Block fromMap)
+    {
+        // bind ConnectorSession if needed
+        if (keyCastFunction.type().parameterArray()[0] == ConnectorSession.class) {
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MapToMapCast.java]: Technically, any argument could be a ConnectorSession

Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MapToMapCast.java]: IIRC, we required functions to put ConnectorSession in the first arg.  If we want to relax that, we'll need to search the code base for other uses of `ConnectorSession.class`

Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MapToMapCast.java]: Most places allow any argument to be of type ConnectorSession (they just loop over the arguments, check for that class and skip or act accordingly). There might be a few places that expect it to be the first arg. We should clean those up.

**Comment by geraint0923** [comment]
Address the comment, I now loop over the parameters to insert `ConnectorSession` if needed.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MapToMapCast.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+                }
+                catch (Throwable t) {
+                    Throwables.propagateIfInstanceOf(t, Error.class);
+                    Throwables.propagateIfInstanceOf(t, PrestoException.class);
+                    throw new PrestoException(StandardErrorCode.INTERNAL_ERROR, t);
+                }
+            }
+            else {
+                // if there are duplicated keys, fail it!
+                throw new PrestoException(StandardErrorCode.INVALID_CAST_ARGUMENT, "duplicated keys found after casting");
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MapToMapCast.java]: "duplicate keys"

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MapToMapCast.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        }
+
+        TypedSet typedSet = new TypedSet(toKeyType, fromMap.getPositionCount() / 2);
+        BlockBuilder keyBlockBuilder = toKeyType.createBlockBuilder(new BlockBuilderStatus(), fromMap.getPositionCount() / 2);
+        for (int i = 0; i < fromMap.getPositionCount(); i += 2) {
+            Object fromKey = readNativeValue(fromKeyType, fromMap, i);
+            try {
+                Object toKey = keyCastFunction.invoke(fromKey);
+                if (toKey == null) {
+                    throw new PrestoException(StandardErrorCode.INVALID_CAST_ARGUMENT, "map key contains null after casting");
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MapToMapCast.java]: "map key is null"

**Comment by martint** [comment]
looks good

**Pull request merged**